crates/new_task/src/dependency.rs

```diff
@@ -0,0 +1,251 @@
+use async_broadcast::Receiver;
+use futures::future::BoxFuture;
+use futures::stream::FuturesUnordered;
+use futures::stream::StreamExt;
+use futures::FutureExt;
+use std::future::Future;
+
+/// Type which describes the idea of waiting for a dependency to complete
+pub trait Dependency<T> {
+    /// Complete will wait until it gets some value `T` then return the value
+    fn completed(self) -> impl Future<Output = T> + Send;
+}
+
+/// Used to combine dependencies to create `AndDependency`s or `OrDependency`s
+trait CombineDependencies<T: Clone + Send + Sync + 'static>:
```

Comment:
Should this be a separate trait? I understand that this has stricter requirements for T, but I don't think there's much utility in using `Dependency` without combinators anyway.

Reply:
I made this a trait so that I could implement it for all of our dependency types, we had, but this is a good point we can promote to the `Dependency` trait.